From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ben Hutchings <ben@decadent.org.uk>
Date: Wed, 21 Aug 2019 00:32:16 +0100
Subject: intel-iommu: Add Kconfig option to exclude iGPU by default

There is still laptop firmware that touches the integrated GPU behind
the operating system's back, and doesn't say so in the RMRR table.
Enabling the IOMMU for all devices causes breakage.

Replace CONFIG_INTEL_IOMMU_DEFAULT_ON with a 3-way choice
corresponding to "on", "off", and "on,intgpu_off".

Signed-off-by: Ben Hutchings <ben@decadent.org.uk>

Bug-Debian: https://bugs.debian.org/935270
Bug-Kali: https://bugs.kali.org/view.php?id=5644

Debian-patch: debian/patches/features/x86/intel-iommu-add-kconfig-option-to-exclude-igpu-by-default.patch (Debian linux 7.2.9-1, salsa.debian.org/kernel-team/linux tag debian/7.2.9-1)
[kernel-soul: rebased on v7.3-rc6. The driver now keeps its state in dmar_policy (DMAR_ON,
 DMAR_DEFAULT_OFF) instead of dmar_disabled, so the first and last hunks of iommu.c were
 rewritten for it. Behaviour is unchanged: the policy is DMAR_ON unless
 INTEL_IOMMU_DEFAULT_OFF is chosen, and dmar_map_intgpu starts at 1 only for
 INTEL_IOMMU_DEFAULT_ON. The Kconfig hunk applies as it was.]
---
 drivers/iommu/intel/Kconfig | 23 +++++++++++++++++------
 drivers/iommu/intel/iommu.c |  9 +++++----
 2 files changed, 22 insertions(+), 10 deletions(-)

diff --git a/drivers/iommu/intel/Kconfig b/drivers/iommu/intel/Kconfig
index 5471f814e0731c4c9f423c6c44aa4b2801782d9b..8263228793a983a8b59bb86a9937da0d71d4b61f 100644
--- a/drivers/iommu/intel/Kconfig
+++ b/drivers/iommu/intel/Kconfig
@@ -60,13 +60,24 @@ config INTEL_IOMMU_SVM
 	  to access DMA resources through process address space by
 	  means of a Process Address Space ID (PASID).
 
-config INTEL_IOMMU_DEFAULT_ON
-	bool "Enable Intel DMA Remapping Devices by default"
-	default y
+choice
+	prompt "Default state of Intel DMA Remapping Devices"
+	default INTEL_IOMMU_DEFAULT_ON
 	help
-	  Selecting this option will enable a DMAR device at boot time if
-	  one is found. If this option is not selected, DMAR support can
-	  be enabled by passing intel_iommu=on to the kernel.
+	  Choose whether Intel DMA Remapping Devices should be enabled
+	  by default. This can be overridden at boot time using the
+	  intel_iommu= kernel parameter.
+
+config INTEL_IOMMU_DEFAULT_ON
+	bool "Enable"
+
+config INTEL_IOMMU_DEFAULT_ON_INTGPU_OFF
+	bool "Enable, excluding integrated GPU"
+
+config INTEL_IOMMU_DEFAULT_OFF
+	bool "Disable"
+
+endchoice
 
 config INTEL_IOMMU_FLOPPY_WA
 	def_bool y
diff --git a/drivers/iommu/intel/iommu.c b/drivers/iommu/intel/iommu.c
index a6007caa8403ae8659bef20076c546bcaf2cc668..ef2e038f4ed355e4e019bc0b7acd9600f5a43773 100644
--- a/drivers/iommu/intel/iommu.c
+++ b/drivers/iommu/intel/iommu.c
@@ -200,17 +200,17 @@ static LIST_HEAD(dmar_satc_units);
 
 static void intel_iommu_domain_free(struct iommu_domain *domain);
 
-#ifdef CONFIG_INTEL_IOMMU_DEFAULT_ON
-int dmar_policy = DMAR_ON;
-#else
+#ifdef CONFIG_INTEL_IOMMU_DEFAULT_OFF
 int dmar_policy = DMAR_DEFAULT_OFF;
+#else
+int dmar_policy = DMAR_ON;
 #endif
 int intel_iommu_sm = IS_ENABLED(CONFIG_INTEL_IOMMU_SCALABLE_MODE_DEFAULT_ON);
 
 int intel_iommu_enabled = 0;
 EXPORT_SYMBOL_GPL(intel_iommu_enabled);
 
-static int dmar_map_intgpu = 1;
+static int dmar_map_intgpu = IS_ENABLED(CONFIG_INTEL_IOMMU_DEFAULT_ON);
 static int intel_iommu_superpage = 1;
 static int iommu_identity_mapping;
 static int iommu_skip_te_disable;
@@ -248,6 +248,7 @@ static int __init intel_iommu_setup(char *str)
 	while (*str) {
 		if (!strncmp(str, "on", 2)) {
 			dmar_policy = DMAR_ON;
+			dmar_map_intgpu = 1;
 			pr_info("IOMMU enabled\n");
 		} else if (!strncmp(str, "off", 3)) {
 			dmar_policy = DMAR_USER_OFF;
